Neos.TimeableNodeVisibility/Classes/Service/TimeableNodeVisibilityService.php

```diff
@@ -0,0 +1,180 @@
+use Neos\ContentRepository\Core\NodeType\NodeTypeNames;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Filter\PropertyValue\Criteria\PropertyValueLessThanOrEqual;
+use Neos\ContentRepository\Core\SharedModel\Node\PropertyName;
+use Neos\ContentRepository\Core\SharedModel\Node\NodeVariantSelectionStrategy;
+use Neos\ContentRepository\Core\Feature\NodeDisabling\Command\EnableNodeAggregate;
+use Psr\Log\LoggerInterface;
+use Neos\ContentRepository\Core\Projection\Workspace\Workspace;
+use Neos\ContentRepository\Core\ContentRepository;
+use Neos\ContentRepository\Core\Feature\NodeDisabling\Command\DisableNodeAggregate;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Filter\PropertyValue\Criteria\OrCriteria;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Node;
+use Neos\ContentRepository\Core\Projection\ProjectionStateInterface;
+use Neos\TimeableNodeVisibility\Domain\HandlingResult;
+use Neos\TimeableNodeVisibility\Domain\HandlingResultSet;
+use Neos\ContentRepository\Core\SharedModel\Exception\WorkspaceDoesNotExist;
+use Neos\ContentRepository\Core\SharedModel\Exception\RootNodeAggregateDoesNotExist;
+
+#[Flow\Scope('singleton')]
+class TimeableNodeVisibilityService
+{
+    #[Flow\Inject]
+    protected ContentRepositoryRegistry $contentRepositoryRegistry;
+
+    #[Flow\Inject]
+    protected LoggerInterface $logger;
+
+    public function handleExceededNodeDates(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName): HandlingResultSet
+    {
+        $contentRepository = $this->contentRepositoryRegistry->get($contentRepositoryId);
+        $liveWorkspace = $contentRepository->getWorkspaceFinder()->findOneByName($workspaceName);
+        if ($liveWorkspace === null) {
+            throw WorkspaceDoesNotExist::butWasSupposedTo($workspaceName);
+        }
+        $nodeHiddenStateFinder = $contentRepository->projectionState(NodeHiddenStateFinder::class);
+
+        $now = new \DateTimeImmutable();
+
+        $nodes = $this->getNodesWithExceededDates($contentRepository, $liveWorkspace, $now);
+        $handledNodeResults = new HandlingResultSet();
+
+        foreach ($nodes as $node) {
+            if ($this->needsEnabling($node, $now) && $this->isHidden($node, $nodeHiddenStateFinder)) {
+                $contentRepository->handle(
+                    EnableNodeAggregate::create(
+                        $node->subgraphIdentity->contentStreamId,
+                        $node->nodeAggregateId,
+                        $node->subgraphIdentity->dimensionSpacePoint,
+                        NodeVariantSelectionStrategy::STRATEGY_ALL_SPECIALIZATIONS
+                    )
+                );
+
+                $result = HandlingResult::createWithEnabled($node);
+                $handledNodeResults->add($result);
+                $this->logResult($result);
+
+            } elseif ($this->needsDisabling($node, $now) && !$this->isHidden($node, $nodeHiddenStateFinder)) {
```

Comment:
as these conditions are mutually exclusive I don't even think we need elseif, can be two separate if statements and we could pull the hidden state right at the beginning and reuse it in both conditions

Reply:
Not sure if this is more performant, as we than fetch the hiddenState always, even `needsEnabling` AND `needsDisabling` is false.

Reply:
Fair point, could be an early return, but I get that might not make things all that much easier.